In [10]:
print("Lima baris pertama dari DataFrame df:")
print(df.head())

print("\nDimensi X_train:", X_train.shape)
print("Dimensi y_train:", y_train.shape)
print("Dimensi X_test:", X_test.shape)
print("Dimensi y_test:", y_test.shape)

Lima baris pertama dari DataFrame df:
        Nama  Gula_g  LemakJenuh_g  Serat_g  Protein_g  Sodium_mg  Kategori  \
0       Apel    10.0           0.0      2.4        0.3          1         1   
1    Brokoli     1.7           0.1      2.6        2.8         33         1   
2  Dada Ayam     0.0           1.0      0.0       31.0         74         1   
3     Salmon     0.0           2.5      0.0       20.0         59         1   
4      Bayam     0.4           0.1      2.2        2.9         79         1   

  Kategori_Nama  
0         Sehat  
1         Sehat  
2         Sehat  
3         Sehat  
4         Sehat  

Dimensi X_train: (11, 5)
Dimensi y_train: (11,)
Dimensi X_test: (5, 5)
Dimensi y_test: (5,)


In [3]:
from sklearn.ensemble import RandomForestClassifier

# Instantiate the RandomForestClassifier model
# Using n_estimators=100 for a robust model
model_rf = RandomForestClassifier(n_estimators=100, random_state=42)

# Train the Random Forest model
model_rf.fit(X_train, y_train)

print("Random Forest model has been successfully trained.")

Random Forest model has been successfully trained.


In [4]:
from sklearn.metrics import accuracy_score, confusion_matrix

# Make predictions on the test set
y_pred = model_rf.predict(X_test)

# Calculate accuracy
akurasi = accuracy_score(y_test, y_pred)
print(f"Akurasi model Random Forest: {akurasi:.2f}")

# Generate confusion matrix
cm = confusion_matrix(y_test, y_pred)
print("\nConfusion Matrix:")
print(cm)

Akurasi model Random Forest: 0.80

Confusion Matrix:
[[2 1]
 [0 2]]


In [5]:
import pandas as pd

# Create a DataFrame for feature importance
feature_importance_df = pd.DataFrame({
    'Fitur': X.columns,
    'Pentingnya': model_rf.feature_importances_
})

# Sort the DataFrame by importance in descending order
feature_importance_df = feature_importance_df.sort_values(by='Pentingnya', ascending=False)

# Display the feature importance DataFrame
print("Pentingnya Fitur dalam Klasifikasi:")
print(feature_importance_df)

Pentingnya Fitur dalam Klasifikasi:
          Fitur  Pentingnya
2       Serat_g    0.246398
4     Sodium_mg    0.239012
0        Gula_g    0.211254
1  LemakJenuh_g    0.205442
3     Protein_g    0.097893


In [6]:
import pandas as pd

# 1. Create a DataFrame for new food data
data_baru = pd.DataFrame({
    'Gula_g': [4, 2],
    'LemakJenuh_g': [2.5, 7.0],
    'Serat_g': [0, 1],
    'Protein_g': [10, 5],
    'Sodium_mg': [35, 1100]
}, index=['Yogurt Yunani', 'Mi Instan'])

print("DataFrame data_baru created:")
print(data_baru.head())

DataFrame data_baru created:
               Gula_g  LemakJenuh_g  Serat_g  Protein_g  Sodium_mg
Yogurt Yunani       4           2.5        0         10         35
Mi Instan           2           7.0        1          5       1100


In [7]:
prediksi_baru = model_rf.predict(data_baru)
print("Predicted categories for new food items:")
print(prediksi_baru)

Predicted categories for new food items:
[1 0]


In [8]:
probabilitas_baru = model_rf.predict_proba(data_baru)
print("Predicted probabilities for new food items (Tidak Sehat, Sehat):")
print(probabilitas_baru)

Predicted probabilities for new food items (Tidak Sehat, Sehat):
[[0.39 0.61]
 [0.82 0.18]]


In [9]:
data_baru['Prediksi'] = prediksi_baru
data_baru['Prediksi_Nama'] = data_baru['Prediksi'].map({0: 'Tidak Sehat', 1: 'Sehat'})
data_baru['Prob_Tidak_Sehat'] = probabilitas_baru[:, 0]
data_baru['Prob_Sehat'] = probabilitas_baru[:, 1]

print("DataFrame data_baru with predictions and probabilities:")
print(data_baru.head())

DataFrame data_baru with predictions and probabilities:
               Gula_g  LemakJenuh_g  Serat_g  Protein_g  Sodium_mg  Prediksi  \
Yogurt Yunani       4           2.5        0         10         35         1   
Mi Instan           2           7.0        1          5       1100         0   

              Prediksi_Nama  Prob_Tidak_Sehat  Prob_Sehat  
Yogurt Yunani         Sehat              0.39        0.61  
Mi Instan       Tidak Sehat              0.82        0.18  


## Penjelasan Sederhana Random Forest

Random Forest adalah algoritma pembelajaran mesin yang sangat kuat dan serbaguna, terutama untuk tugas klasifikasi dan regresi. Nama 'Random Forest' sendiri memberikan petunjuk tentang cara kerjanya: ini adalah 'hutan' yang terdiri dari banyak 'pohon' acak.

### 1. Konsep Dasar: Hutan Pohon Keputusan

Bayangkan Anda memiliki beberapa teman yang ahli dalam memprediksi apakah suatu makanan sehat atau tidak sehat. Setiap teman ini (kita sebut saja 'pohon keputusan') akan melihat beberapa ciri makanan (seperti jumlah gula, lemak jenuh, serat, dll.) dan kemudian membuat keputusan sendiri. Misalnya, satu teman mungkin sangat memperhatikan gula, sementara yang lain lebih fokus pada serat dan sodium.

Random Forest bekerja dengan membangun banyak (biasanya ratusan atau ribuan) pohon keputusan yang berbeda secara independen. Setiap pohon dilatih menggunakan sebagian kecil data pelatihan yang dipilih secara acak (disebut bootstrapping) dan hanya mempertimbangkan sebagian kecil fitur secara acak pada setiap 'cabang' keputusan. Keragaman ini membuat setiap pohon memiliki perspektif yang sedikit berbeda tentang data.

### 2. Mengapa Random Forest Lebih Baik dari Satu Pohon Keputusan?

Satu pohon keputusan tunggal cenderung 'overfit' pada data pelatihan. Artinya, pohon tersebut mungkin terlalu spesifik mengingat detail kecil dari data pelatihan sehingga tidak dapat menggeneralisasi dengan baik ke data baru yang belum pernah dilihat sebelumnya. Ini seperti seorang teman yang menghafal semua jawaban ujian tanpa benar-benar memahami konsepnya.

Random Forest mengatasi masalah ini dengan menggabungkan hasil dari banyak pohon. Ini mirip dengan meminta pendapat dari banyak ahli yang berbeda. Jika satu ahli membuat kesalahan, kesalahan tersebut kemungkinan akan diimbangi oleh keputusan yang benar dari ahli lainnya. Ini membuat model lebih stabil dan akurat, serta sangat mengurangi risiko overfitting.

### 3. Bagaimana Random Forest Membuat Keputusan Akhir (Voting)?

Setelah semua pohon keputusan dilatih, ketika Random Forest menerima data baru (misalnya, data nutrisi makanan baru), setiap pohon akan membuat prediksinya sendiri (misalnya, 'sehat' atau 'tidak sehat').

Untuk mendapatkan keputusan akhir, Random Forest menggunakan mekanisme 'voting' (pemungutan suara). Untuk masalah klasifikasi seperti kita, prediksi akhir adalah kelas yang paling banyak dipilih oleh pohon-pohon individu. Jadi, jika 70 dari 100 pohon memprediksi 'Sehat' dan 30 memprediksi 'Tidak Sehat', maka Random Forest akan secara keseluruhan memprediksi 'Sehat'.

### 4. Kaitannya dengan Klasifikasi Makanan Sehat/Tidak Sehat

Dalam kasus klasifikasi makanan sehat/tidak sehat yang telah kita lakukan, Random Forest menggunakan fitur-fitur nutrisi seperti `Gula_g`, `LemakJenuh_g`, `Serat_g`, `Protein_g`, dan `Sodium_mg` sebagai input. Setiap pohon keputusan dalam 'hutan' akan menimbang fitur-fitur ini secara berbeda untuk membuat prediksinya sendiri. Misalnya:

*   Pohon 1 mungkin memprioritaskan `Serat_g` dan `Gula_g`.
*   Pohon 2 mungkin lebih fokus pada `LemakJenuh_g` dan `Sodium_mg`.

Setelah semua pohon 'memberi suara' apakah makanan baru itu 'Sehat' (1) atau 'Tidak Sehat' (0), Random Forest akan menghitung suara tersebut. Makanan akan diklasifikasikan ke dalam kategori yang menerima suara terbanyak. Probabilitas yang kita lihat (`probabilitas_baru`) adalah persentase suara untuk setiap kategori. Misalnya, `[0.39 0.61]` berarti 39% pohon memilih 'Tidak Sehat' dan 61% memilih 'Sehat'.